classes/controlledVocab/Repository.php

```diff
@@ -0,0 +1,199 @@
+        $entryDao = $entryDaoClassName
+            ? DAORegistry::getDAO($entryDaoClassName)
+            : $this->getEntryDaoBySymbolic($symbolic);
+
+        $currentControlledVocab = $this->build($symbolic, $assocType, $assocId);
+
+        if ($deleteFirst) {
+            collect($currentControlledVocab->enumerate( $symbolic))
+                ->keys()
+                ->each(fn (int $id) => $entryDao->deleteObjectById($id));
+        }
+
+        if (is_array($vocabs)) { // localized, array of arrays
+            foreach ($vocabs as $locale => $list) {
+                if (is_array($list)) {
+                    $list = array_unique($list); // Remove any duplicate keywords
+                    $i = 1;
+                    foreach ($list as $vocab) {
+                        $vocabEntry = $entryDao->newDataObject();
+                        $vocabEntry->setControlledVocabId($currentControlledVocab->id);
+                        $vocabEntry->setData($symbolic, $vocab, $locale);
+                        $vocabEntry->setSequence($i);
+                        $i++;
+                        $entryDao->insertObject($vocabEntry);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Update a user's set of interests
+     */
+    public function setUserInterests(array $interests, int $userId): void
+    {
+        $controlledVocab = Repo::controlledVocab()->build(
+            UserInterest::CONTROLLED_VOCAB_INTEREST
+        );
+
+        /** @var InterestEntryDAO $interestEntryDao */
+        $interestEntryDao = DAORegistry::getDAO('InterestEntryDAO');
+
+        DB::beginTransaction();
+
+        // Delete the existing interests association.
+        UserInterest::withUserId($userId)->delete();
+
+        collect($interests)
+            ->map(fn (string $interest): string => trim($interest))
+            ->unique()
+            ->each(function (string $interest) use ($controlledVocab, $interestEntryDao, $userId): void {
+                $interestEntry = $interestEntryDao->getBySetting(
+                    $interest,
+                    $controlledVocab->symbolic,
+                    $controlledVocab->assocId,
+                    $controlledVocab->assocType,
+                    $controlledVocab->symbolic
+                );
+
+                if (!$interestEntry) {
+                    $interestEntry = $interestEntryDao->newDataObject(); /** @var InterestEntry $interestEntry */
+                    $interestEntry->setInterest($interest);
+                    $interestEntry->setControlledVocabId($controlledVocab->id);
+                    $interestEntry->setId($interestEntryDao->insertObject($interestEntry));
+                }
+
+                UserInterest::create([
+                    'userId' => $userId,
+                    'controlledVocabEntryId' => $interestEntry->getId(),
+                ]);
+            });
+        
+        DB::commit();
+    }
```

Comment:
perhaps this one can be moved to `UserInterest` model as a static method but seems kind of too much to have for a model . Then again this also seems kind of out of place in the vocab repo but then again it does handle a specific type of vocab .

Reply:
Agreed, it shouldn't be here. I'd prefer `UserInterest` as a static method, or perhaps better a `Repository` class alongside `UserInterest`.